File: roller.c
// Author: Andreas Fladstad
// Compile with gcc roller.c -o roller

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int main(int argc, char** argv) {
  if(argc != 2) {
    printf("Usage: ./%s <roll>", argv[0]);
    return -1;
  }

  // Parse dice

  int amount = 0;
  int dice = 0;
  int plus = 0;
  char* input = argv[1];
  char buffer[20];
  int count = 0;

  while(*input != 'd') {
    if(*input == 0) {
      printf("error in input\n");
    }
    buffer[count++] = *input;
    input++;
  }
  input++;
  buffer[count] = 0;

  if(count != 0) {
    amount = atoi(buffer);
  }

  count = 0;

  while(*input != '+' && *input != 0) {
    buffer[count++] = *input;
    input++;
  }

  buffer[count] = 0;
  dice = atoi(buffer);

  if(*input == '+') {
    count = 0;
    input++;
    while(*input != 0) {
      buffer[count++] = *input;
      input++;
    }
    buffer[count] = 0;
    plus = atoi(buffer);
  }

  // Roll dice
  srand(time(NULL));
  int result = 0;
  for(int i = 0; i < amount; i++) {
    int current = rand() % dice + 1;
    result += current;
    printf("%d ", current);
  }
  if(plus != 0) {
    printf("+ %d", plus);
  }
  printf(" = %d\n", result);
  return 0;
}

File: roller.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"regexp"
	"strconv"
	"time"
)

func main() {
	if len(os.Args) != 2 {
		fmt.Printf("Usage ./%s (<amount>)d<dice>(+<plus>)")
	}
	reg := regexp.MustCompile(`(?P<Amount>\d*)?d(?P<Dice>\d*)\+?(?P<Plus>\d*)?`)
	input := os.Args[1]
	var (
		dice    = 0
		amount  = 1
		plus    = 0
		matches = reg.FindStringSubmatch(input)
	)

	paramsMap := make(map[string]string)
	for i, name := range reg.SubexpNames() {
		paramsMap[name] = matches[i]
	}

	if paramsMap["Plus"] != "" {
		num, err := strconv.Atoi(paramsMap["Plus"])
		if err == nil {
			plus = num
		}
	}

	if paramsMap["Dice"] != "" {
		num, err := strconv.Atoi(paramsMap["Dice"])
		if err == nil {
			dice = num
		}
	}

	if paramsMap["Amount"] != "" {
		num, err := strconv.Atoi(paramsMap["Amount"])
		if err == nil {
			amount = num
		}
	}

	result := 0
	source := rand.NewSource(time.Now().UnixNano())
	r := rand.New(source)

	for i := 0; i < amount; i++ {
		current := r.Intn(dice) + 1
		result += current
		fmt.Printf("%d ", current)
	}
	if plus != 0 {
		result += plus
		fmt.Printf("+ %d", plus)
	}
	fmt.Printf(" = %d\n", result)
}
